### PR TITLE
network: fix reference counting issues

### LINKED ISSUE #19629: systemd-networkd randomly crashes in CI (`Assertion 'link->dhcp4_messages > 0' failed at src/network/networkd-dhcp4.c:1013, function int dhcp4_address_handler(sd_netlink *, sd_netlink_message *, Link *)()`)
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > Arch Linux

In a couple of recent runs I noticed that systemd-networkd crashes with SIGABRT somewhere during the CI run without having any impact on test results:

```
arch.localdomain systemd-resolved[121430]: Removing scope on link veth99, protocol llmnr, family AF_INET6
arch.localdomain systemd-networkd[122303]: veth99: Starting IPv4ACD client. Probing DHCPv4 address 192.168.5.10
arch.localdomain systemd-resolved[121430]: Removing scope on link veth-peer, protocol llmnr, family AF_INET
arch.localdomain systemd-resolved[121430]: Removing scope on link veth-peer, protocol llmnr, family AF_INET6
arch.localdomain systemd-udevd[121433]: Network interface NamePolicy= disabled on kernel command line, ignoring.
arch.localdomain systemd-resolved[121430]: Removing link 75/veth99
arch.localdomain systemd-resolved[121430]: Removing link 74/veth-peer
arch.localdomain systemd-networkd[122303]: veth99: Failed to start IPv4ACD for DHCPv4 address: No such device
arch.localdomain systemd-networkd[122303]: veth99: Failed
arch.localdomain systemd-networkd[122303]: veth99: State changed: configuring -> failed
arch.localdomain systemd-networkd[122303]: veth99: DHCPv4 client: STOPPED
arch.localdomain systemd-networkd[122303]: veth99: LLDP: Stopping LLDP client
arch.localdomain systemd-networkd[122303]: veth99: Flags change: -UP -LOWER_UP -RUNNING
arch.localdomain systemd-networkd[122303]: veth99: Link DOWN
arch.localdomain systemd-networkd[122303]: veth99: Lost carrier
arch.localdomain systemd-networkd[122303]: veth99: DHCPv4 client: STOPPED
arch.localdomain systemd-networkd[122303]: veth99: Removing address: 192.168.5.10/24 (valid for 1h, preferred for 1h)
arch.localdomain systemd-networkd[122303]: veth99: State is failed, dropping foreign config
arch.localdomain systemd-networkd[122303]: veth-peer: LLDP: Failed to determine datagram size to read, ignoring: Network is down
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: veth99: Forgetting route: dst: fe80::/64, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main(254), proto: kernel, type: unicast, nexthop: 0, priority: 256
arch.localdomain systemd-networkd[122303]: veth99: Forgetting route: dst: fe80::1034:56ff:fe78:9abc/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0
arch.localdomain systemd-networkd[122303]: veth99: Forgetting route: dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: multicast, nexthop: 0, priority: 256
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: veth99: Forgetting address: fe80::1034:56ff:fe78:9abc/64 (valid forever, preferred forever)
arch.localdomain systemd-networkd[122303]: veth-peer: Flags change: -UP -LOWER_UP -RUNNING
arch.localdomain systemd-networkd[122303]: veth-peer: Link DOWN
arch.localdomain systemd-networkd[122303]: veth-peer: LLDP: Stopping LLDP client
arch.localdomain systemd-networkd[122303]: veth-peer: Stopped LLDP.
arch.localdomain systemd-networkd[122303]: veth-peer: Lost carrier
arch.localdomain systemd-networkd[122303]: veth-peer: DHCPv4 server: STOPPED
arch.localdomain systemd-networkd[122303]: veth-peer: Removing address: 192.168.5.1/24 (valid forever, preferred forever)
arch.localdomain systemd-networkd[122303]: veth-peer: Removing address: 192.168.5.10/24 (valid forever, preferred forever)
arch.localdomain systemd-networkd[122303]: veth-peer: State is configured, dropping foreign config
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting route: dst: fe80::/64, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main(254), proto: kernel, type: unicast, nexthop: 0, priority: 256
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting route: dst: fe80::1034:56ff:fe78:9abd/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting route: dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: multicast, nexthop: 0, priority: 256
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: rtnl: received non-static neighbor, ignoring.
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting address: fe80::1034:56ff:fe78:9abd/64 (valid forever, preferred forever)
arch.localdomain systemd-networkd[122303]: veth99: State changed: failed -> linger
arch.localdomain systemd-networkd[122303]: veth99: Link removed
arch.localdomain systemd-networkd[122303]: veth99: netdev removed
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting address: 192.168.5.1/24 (valid forever, preferred forever)
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting route: dst: 192.168.5.1/32, src: n/a, gw: n/a, prefsrc: 192.168.5.1, scope: host, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0
arch.localdomain systemd-networkd[122303]: veth-peer: Forgetting address: 192.168.5.10/24 (valid forever, preferred forever)
arch.localdomain systemd-networkd[122303]: veth-peer: State changed: configured -> linger
arch.localdomain systemd-networkd[122303]: veth-peer: Link removed
arch.localdomain systemd-networkd[122303]: Assertion 'link->dhcp4_messages > 0' failed at src/network/networkd-dhcp4.c:1013, function int dhcp4_address_handler(sd_netlink *, sd_netlink_message *, Link *)(). Aborting.
arch.localdomain systemd[1]: Started Process Core Dump (PID 122328/UID 0).
arch.localdomain systemd-coredump[122329]: Resource limits disable core dumping for process 122303 (systemd-network).
arch.localdomain systemd-coredump[122329]: Process 122303 (systemd-network) of user 980 dumped core.
arch.localdomain systemd[1]: systemd-coredump@6-122328-0.service: Deactivated successfully.
arch.localdomain systemd[1]: systemd-networkd.service: Main process exited, code=dumped, status=6/ABRT
arch.localdomain systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
arch.localdomain systemd[1]: Starting Network Service...
```

(taken from https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/4679/artifact//systemd-centos-ci/artifacts_all/artifacts_EcYd8l/vagrant-logs.NVD/vagrant-arch-sanitizers-clang-testsuite.cRh/journalctl-testsuite_PASS.log)

/cc @yuwata

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #19629.

### CODE REVIEW

**Comment by yuwata** [comment]
@mrc0mmand I am not sure this fixes #19629 or not. Could you test this? I cannot reproduce the issue on my laptop.

**Comment by mrc0mmand** [comment]
> @mrc0mmand I hope that the first commit fixes the issue #19629. But I am not sure. Could you test this? I cannot reproduce the issue on my laptop.

Thanks for the fix! Unfortunately, I can't reproduce it either, since it seems to be a race condition and so far I've seen it in CI just a couple of times. I guess only time will tell if it indeed works.

**Comment by yuwata** [comment]
> > @mrc0mmand I hope that the first commit fixes the issue #19629. But I am not sure. Could you test this? I cannot reproduce the issue on my laptop.
> 
> Thanks for the fix! Unfortunately, I can't reproduce it either, since it seems to be a race condition and so far I've seen it in CI just a couple of times. I guess only time will tell if it indeed works.

@mrc0mmand OK, thanks.

**Comment by keszybz** [comment]
LGTM.

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -1149,6 +1149,8 @@ int request_process_address(Request *req) {
         if (r < 0)
                 return r;
 
+        req->message_counter = NULL;
```
Comment (bluca) [inline src/network/networkd-address.c:1153]: probably worth having at least a one line comment explaining why this is done, taken from the commit message

**Review by bluca — APPROVED** [review]
lgtm, just one minor nit

File: src/network/networkd-address.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-address.c:1153]: Thanks. Done.

**Pull request merged**